syntax = "proto3";

package sdf;


message Credentials {
  string subscriptionId = 1;
  string resourceGroup = 2;
  string workspaceName = 3;
}

// The details of an existing dataset in the workspace to be used.
message Dataset {
  string trainingFile = 1;
  string name = 2;
  // The name to be used by the entry script to refer to the dataset
  string lookupKey = 3;
}

// The specs for an environment to create with the necessary packages
// TODO: After successful runs, we should register the environments
//       for future re-use
message Environment {
  // Decides if to run on the node or submit to Azure ML workspace
  bool localRun = 1;
  string name = 2;
  string reqsPath = 3;
}

message ExperimentSetup {

  Environment env = 1;

  string experimentName = 2;

  Credentials access = 3;

  Dataset dataset = 4;

  // The next two settings are used in the ScriptRunConfig
  // The compute cluster under which to submit the experiment
  string computeCluster = 5;
  // The script to use as an entry point for the built container
  string entryScript = 6;

}

message ExperimentResult {
  string resultSummary = 1;
}

enum ComputeCallBacks {
  DUMMY_FIELD = 0;
  PROCESS_RESULTS = 1;
}


// The specs for interacting with the sharpening cloud, whether by calling
// NASA or AWS DynamoDB
message Layer {
  string name = 1;
}

// May require rotation in the near future
// For now, they can be loaded as environment vars or ConfigMap
message EarthCloudCredentials {
  string username = 1;
  string password = 2;
}

message EarthCloudProduct {
  string fullVersion = 1;
  string description = 2;
  repeated Layer layers = 3;
  string projection = 4;
}

// Optional for requesting modules
message Request {
  string name = 1;
  bytes specs = 2;
}

message EarthCloudConfig {
  EarthCloudCredentials access = 1;
  repeated EarthCloudProduct products = 2;
  Request request = 3;
}

message EarthClouResult {
  string result = 1;
}
